File: sudoku.py
import random

def print_board(board):
    for i in range(9):
        if i % 3 == 0 and i != 0:
            print("- - - - - - - - - - - -")

        for j in range(9):
            if j % 3 == 0 and j != 0:
                print(" | ", end="")

            if j == 8:
                print(board[i][j])
            else:
                print(str(board[i][j]) + " ", end="")

def is_valid(board, row, col, num):
    # Check if the number is already in the current row
    if num in board[row]:
        return False
    
    # Check if the number is already in the current column
    for i in range(9):
        if board[i][col] == num:
            return False
    
    # Check if the number is already in the current 3x3 grid
    start_row = row - row % 3
    start_col = col - col % 3
    for i in range(3):
        for j in range(3):
            if board[i + start_row][j + start_col] == num:
                return False
    
    return True

def solve_sudoku(board):
    empty_cell = find_empty_cell(board)
    if not empty_cell:
        return True
    
    row, col = empty_cell
    
    for num in range(1, 10):
        if is_valid(board, row, col, num):
            board[row][col] = num
            
            if solve_sudoku(board):
                return True
            
            board[row][col] = 0
    
    return False

def find_empty_cell(board):
    for i in range(9):
        for j in range(9):
            if board[i][j] == 0:
                return (i, j)
    return None

def generate_sudoku():
    board = [[0 for _ in range(9)] for _ in range(9)]
    solve_sudoku(board)
    # Remove some numbers to create a puzzle
    for _ in range(45):  # Adjust the number of empty cells for desired difficulty
        row = random.randint(0, 8)
        col = random.randint(0, 8)
        board[row][col] = 0
    return board

if __name__ == "__main__":
    sudoku_board = generate_sudoku()
    print("Generated Sudoku Puzzle:")
    print_board(sudoku_board)

File: sudoku.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define SIZE 9

void print_board(char board[SIZE][SIZE]) {
    for (int i = 0; i < SIZE; i++) {
        if (i % 3 == 0 && i != 0) {
            printf("- - - - - - - - - - - -\n");
        }

        for (int j = 0; j < SIZE; j++) {
            if (j % 3 == 0 && j != 0) {
                printf(" | ");
            }

            if (j == SIZE - 1) {
                printf("%c\n", board[i][j]);
            } else {
                printf("%c ", board[i][j]);
            }
        }
    }
}

int is_valid(char board[SIZE][SIZE], int row, int col, char num) {
    for (int i = 0; i < SIZE; i++) {
        if (board[row][i] == num) {
            return 0;
        }
    }

    for (int i = 0; i < SIZE; i++) {
        if (board[i][col] == num) {
            return 0;
        }
    }

    int start_row = row - row % 3;
    int start_col = col - col % 3;
    for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 3; j++) {
            if (board[i + start_row][j + start_col] == num) {
                return 0;
            }
        }
    }

    return 1;
}

int find_empty_cell(char board[SIZE][SIZE], int *row, int *col) {
    for (int i = 0; i < SIZE; i++) {
        for (int j = 0; j < SIZE; j++) {
            if (board[i][j] == '_') {
                *row = i;
                *col = j;
                return 1;
            }
        }
    }
    return 0;
}

int solve_sudoku(char board[SIZE][SIZE]) {
    int row, col;

    if (!find_empty_cell(board, &row, &col)) {
        return 1;
    }

    for (char num = '1'; num <= '9'; num++) {
        if (is_valid(board, row, col, num)) {
            board[row][col] = num;

            if (solve_sudoku(board)) {
                return 1;
            }

            board[row][col] = '_';
        }
    }

    return 0;
}

void generate_sudoku(char board[SIZE][SIZE]) {
    for (int i = 0; i < SIZE; i++) {
        for (int j = 0; j < SIZE; j++) {
            board[i][j] = '_';
        }
    }

    solve_sudoku(board);

    for (int i = 0; i < 45; i++) {
        int row = rand() % SIZE;
        int col = rand() % SIZE;
        board[row][col] = '_';
    }
}

void play_sudoku(char board[SIZE][SIZE]) {
    int row, col;
    char num;

    while (1) {
        printf("Enter row (1-9), column (1-9), and number (1-9) separated by spaces (enter 0 0 0 to exit): ");
        scanf("%d %d %c", &row, &col, &num);

        if (row == 0 && col == 0 && num == '0') {
            break;
        }

        row -= 1;
        col -= 1;

        if (row >= 0 && row < SIZE && col >= 0 && col < SIZE && num >= '1' && num <= '9') {
            if (board[row][col] == '_') {
                if (is_valid(board, row, col, num)) {
                    board[row][col] = num;
                    print_board(board);
                } else {
                    printf("Invalid move!\n");
                }
            } else {
                printf("Cell is already filled!\n");
            }
        } else {
            printf("Invalid input!\n");
        }
    }
}

int main() {
    srand(time(0));

    char sudoku_board[SIZE][SIZE];
    generate_sudoku(sudoku_board);

    printf("Generated Sudoku Puzzle:\n");
    print_board(sudoku_board);

    play_sudoku(sudoku_board);

    return 0;
}
